Mathlib/Lean/Meta/Compare.lean

```diff
@@ -0,0 +1,287 @@
+  compareMetavarDecls? : Option MetavarDeclComparisonConfig := some {}
+deriving Repr, Inhabited
+
+/-! ## Preset Configs -/
+
+/-! ### Any changes -/
+
+/-- Check for any changes whatsoever. -/
+def LocalDeclComparisonConfig.anyChanges : LocalDeclComparisonConfig where
+  checkIndex         := true
+  checkFVarId        := true
+  checkLocalDeclKind := true
+
+/-- Check for any changes whatsoever. -/
+def LocalContextComparisonConfig.anyChanges : LocalContextComparisonConfig where
+  includeAuxDecls    := true
+  includeImplDetails := true
+  toLocalDeclComparisonConfig := .anyChanges
+
+/-- Check for any changes whatsoever. -/
+def MetavarDeclComparisonConfig.anyChanges : MetavarDeclComparisonConfig where
+  checkIndex := true
+  compareLocalContexts? := some .anyChanges
+
+/-- Check for any changes whatsoever. -/
+def MVarIdComparisonConfig.anyChanges : MVarIdComparisonConfig where
+  checkMVarId := true
+  compareMetavarDecls? := some .anyChanges
+
+/-! ### Only expressions -/
+
+/-- Only compare expressions appearing in a local decl. -/
+def LocalDeclComparisonConfig.onlyExprs : LocalDeclComparisonConfig where
+  checkUserName   := false
+  checkBinderInfo := false
+
+/-- Only compare expressions appearing in the local context. (Checks local instances.) -/
+def LocalContextComparisonConfig.onlyExprs : LocalContextComparisonConfig where
+  toLocalDeclComparisonConfig := .onlyExprs
+
+/-- Only compare expressions appearing in the target and local context. (Does not include
+implementation details; does check local instances.) -/
+def MetavarDeclComparisonConfig.onlyExprs : MetavarDeclComparisonConfig where
+  checkUserName    := false
+  checkMetavarKind := false
+  compareLocalContexts? := some .onlyExprs
+
+/-- Only compare expressions appearing in the target and local context. (Does not include
+implementation details; does check local instances.) -/
+def MVarIdComparisonConfig.onlyExprs : MVarIdComparisonConfig where
+  compareMetavarDecls? := some .onlyExprs
+
+end Mathlib.Meta
+
+/-! # Comparisons -/
+
+namespace Lean
+
+open Lean Meta Mathlib.Meta
+
+/-- Compares two expressions according to the given `ExprComparisonConfig`. See the documentation
+for `ExprComparisonConfig` for more information. -/
+def Expr.compare (e₁ e₂ : Expr) (config : ExprComparisonConfig := {}) : MetaM Bool :=
+  match config with
+  | ⟨.beq, _⟩ => pure (e₁ == e₂)
+  | ⟨.defEq, t⟩ => withTransparency t <| withNewMCtxDepth <| isDefEq e₁ e₂
+
+/-- Compares two lists monadically. Does not check length first. Unrelated to `ListM`. -/
+private def compareListM (eq : α → α → MetaM Bool) (l₁ l₂ : List α) : MetaM Bool :=
+  l₁.zip l₂ |>.allM fun (a, b) => eq a b
+
+/-- Compare two `LocalDecl`s according to the configs. The settings in `ExprComparisonConfig` will
+be used when comparing expressions. -/
+def LocalDecl.compare (l₁ l₂ : LocalDecl) (cfg : LocalDeclComparisonConfig := {})
+    (ecfg : ExprComparisonConfig) : MetaM Bool :=
```

Comment:
There was a reason for this—I think. I really ought to have documented it at the time. :) I'll try to remind myself about it tomorrow when I'm feeling a bit more refreshed.

Reply:
Ah, right. The reason was that we use the following pattern for specifying e.g. the `LocalContextComparisonConfig` in the `MetavarDeclComparisonConfig`:
```
structure MetavarDeclComparisonConfig where
  ...
  compareLocalContexts? : Option LocalContextComparisonConfig := some {}
```
If we don't want to compare the local contexts at all, we set this to `none`.

If `LocalContextComparisonConfig` extended `ExprComparisonConfig`, then its default values would be different than the "global" values. This would cause unexpected behavior: you set `.eqKind := `.isDefEq` at the top level, but don't wind up using `.isDefEq` on the local context!

We could get around this in two other ways if we wanted to:
1. by having `LocalContextComparisonConfig` take in a parameter of type `ExprComparisonConfig` which was set by other fields of `MetavarDeclComparisonConfig` and only existed to propagate the "global" defaults in from `MetavarDeclComparisonConfig`. This was advised against here.
2. by using a different pattern: have a field `checkLocalContext : Bool` in `MetavarDeclComparisonConfig` which toggled whether the comparison happened, and having `MetavarDeclComparisonConfig` extend `LocalContextComparisonConfig`. I thought this might be a little more error-prone, as then you'd still have all the `LocalContextComparisonConfig` fields floating around even when they weren't being used at all. By using an `Option` we prevent this data from even existing when it's not relevant.

I'll add this to the module doc, unless we want to change the way it works. :)

Reply:
Thanks for the explanation! I think that's a good argument to leave it as it!